aws/terminator/compute.py

```diff
@@ -618,3 +618,82 @@ def created_time(self):
 
     def terminate(self):
         self.client.cancel_spot_instance_requests(SpotInstanceRequestIds=[self.id])
+
+
+class Ecs(Terminator):
+    @staticmethod
+    def create(credentials):
+        def _paginate_cluster_results(client):
+            names = client.get_paginator('list_clusters').paginate(
+                PaginationConfig={
+                    'PageSize': 100,
+                }
+            ).build_full_result()['clusterArns']
+
+            if not names:
+                return []
+
+            return client.describe_clusters(name=names)['clusters']
+
+        return Terminator._create(credentials, Ecs, 'ecs', _paginate_cluster_results)
+
+    @property
+    def name(self):
+        return self.instance['clusterName']
+
+    def terminate(self):
+        def _paginate_task_results(container_instance):
+            names = self.client.get_paginator('list_tasks').paginate(
+                cluster=self.name,
+                containerInstance=container_instance,
+                PaginationConfig={
+                    'PageSize': 100,
+                }
+            ).build_full_result()['taskArns']
+
+            if not names:
+                return []
+
+            return self.client.describe_taks(cluster=self.name, tasks=names)['tasks']
+
+        def _paginate_container_instance_results():
+            names = self.client.get_paginator('list_container_instances').paginate(
+                cluster=self.name,
+                PaginationConfig={
+                    'PageSize': 100,
+                }
+            ).build_full_result()['containerInstanceArns']
+
+            if not names:
+                return []
+
+            return self.client.describe_container_instances(cluster=self.name, containerInstances=names)['containerInstances']
+
+        def _paginate_service_results():
+            names = self.client.get_paginator('list_services').paginate(
+                cluster=self.name,
+                PaginationConfig={
+                    'PageSize': 100,
+                }
+            ).build_full_result()['serviceArns']
+
+            if not names:
+                return []
+
+            return self.client.describe_services(cluster=self.name, services=names)['services']
+
+        container_instances = _paginate_container_instance_results()
+        for name in container_instances:
+            # If there are tasks running on a container instance, stop them first
+            tasks = _paginate_task_results(name['containerInstanceArn'])
+            for task in tasks:
+                self.client.stop_task(cluster=self.name, task=task['taskArn'])
```

Comment:
```suggestion
            tasks = _paginate_task_results(name['containerInstanceArn'])
            for task in tasks:
                self.client.stop_task(cluster=self.name, task=task['taskArn'])
```

Unless the desired count of the belonging service is not set  to 0, the service might spawn new tasks to replace the stopped ones.

Reply:
maybe flipping solves this issue in a lazy way?

1. delete the service
2. delete all left running tasks (_like `killall -9`_)
3. delete the cluster

Reply:
Do the tasks need to be individually deleted at all? Won't deregistering the container instance also remove the tasks?

Reply:
If the ecs_tasks runs on ec2, yes. deleteing the instance, kills also the ecs tasks.  
but if it's serverless (fargate), there is no otherway to stop/kill the tasks.  

but basically you did it that way, it you won't wait for the service to become inactive.

imo we've got two choices.  
a) delete service and wait until it's gone. forward delete cluster  
b) delete service, kill pending tasks, forward delete cluster

Reply:
I feel like the first option seems like less work.

Reply:
https://github.com/mattclay/aws-terminator/pull/210#discussion_r879421673

Then we must implement waiters for the service.

Reply:
If a service truly cannot be deleted until all tasks are stopped, I don't think we can use a waiter. A task could take up to 2 minutes to shutdown, plus however long it takes for amazon to issue the shutdown request. We can't block the lambda for that long.

Reply:
The terminator must not wait on operations to complete, as that will block it from performing other actions. Resource age limits should be used to control the order of deletions when dependencies are involved.

Reply:
@gravesm @mattclay another possibility is to wrap `delete_cluster` into try/except.  
imo, the ecs services and tasks will be stopped/deleted. maybe it takes some time until they gone an the `delete_cluster` will fail because of that. but it doesn't matter. because the next run of the terminator will delete the empty cluster then. and the ecs cluster itself has no fees

Reply:
We wouldn't want to ignore the exception forever though.  So perhaps, one class with the default `age_limit` that catches and ignores `ClusterContainsServicesException` and `ClusterContainsTasksException` on `delete_cluster()`.  Then a second class, which has an `age_limit` 10 minutes longer than the default, that only tries to delete clusters (and that one doesn't ignore those exceptions, so that errors still surface to us in the monitoring).

I'm assuming 10 minutes should be more than long enough, even on a bad day, for all tasks and services to be 100% gone.

Reply:
That sounds good!

Reply:
@markuman I tried to create the second class, @markuman can you check please and let me know if this is what you meant? Thanks.

Reply:
LGTM.
What's about the failing CI. I cannot find any usefull error message.

Reply:
@markuman it needs to be refactored in some way because the file's dimension exceeds the limit. Will try to see what I can do for this thing.